src/appengine/handlers/cron/fuzz_strategy_selection.py

```diff
@@ -0,0 +1,154 @@
+
+See https://www.cs.mcgill.ca/~vkules/bandits.pdf
+"""
+MULTI_ARMED_BANDIT_PROB = """
+SELECT
+  EXP(standardized_avg_ne / temp_p) / (SUM(EXP(standardized_avg_ne / temp_p)) OVER()) AS edges_bandit,
+  EXP(standardized_avg_cc / temp_p) / (SUM(EXP(standardized_avg_cc / temp_p)) OVER()) AS crash_bandit,
+  EXP(standardized_avg_nf / temp_p) / (SUM(EXP(standardized_avg_nf / temp_p)) OVER()) AS features_bandit,
+  EXP(standardized_avg_nu / temp_p) / (SUM(EXP(standardized_avg_nu / temp_p)) OVER()) AS units_bandit,
+  strategy,
+  strategy_count
+FROM (
+  SELECT
+    CONCAT(s_radamsa, s_max_len, s_ml_rnn, s_vp, s_corpus, s_fork, s_subset, s_recommended_dict) AS strategy,
+    AVG((crash_count - avg_cc) / stddev_cc) AS standardized_avg_cc,
+    AVG((new_edges - avg_ne) / stddev_ne) AS standardized_avg_ne,
+    AVG((new_features - avg_nf) / stddev_nf) AS standardized_avg_nf,
+    AVG((new_units_added - avg_nu) / stddev_nu) AS standardized_avg_nu,
+    COUNT(*) AS strategy_count,
+    "any" AS matcher,
+    /* Change temperature parameter here. */ .2 AS temp_p
+  FROM (
+    SELECT
+    IF
+      (strategy_corpus_mutations_radamsa > 0,
+        "radamsa,",
+        "") AS s_radamsa,
+    IF
+      (strategy_random_max_len>0,
+        "max len,",
+        "") AS s_max_len,
+    IF
+      (strategy_corpus_mutations_ml_rnn>0,
+        "ml rnn,",
+        "") AS s_ml_rnn,
+    IF
+      (strategy_value_profile > 0,
+        "value profile,",
+        "") AS s_vp,
+    IF
+      (strategy_corpus_mutations > 0,
+        "corpus,",
+        "") AS s_corpus,
+    IF
+      (strategy_fork > 0,
+        "fork,",
+        "") AS s_fork,
+    IF
+      (strategy_corpus_subset > 0,
+        "subset,",
+        "") AS s_subset,
+    IF
+      (strategy_recommended_dict > 0,
+        "dict,",
+        "") AS s_recommended_dict,
+      AVG(crash_count) OVER() AS avg_cc,
+      AVG(new_edges) OVER() AS avg_ne,
+      AVG(new_features) OVER() AS avg_nf,
+      AVG(new_units_added) OVER() AS avg_nu,
+      STDDEV(crash_count) OVER() AS stddev_cc,
+      STDDEV(new_edges) OVER() AS stddev_ne,
+      STDDEV(new_features) OVER() AS stddev_nf,
+      STDDEV(new_units_added) OVER() AS stddev_nu,
+      crash_count,
+      new_edges,
+      new_features,
+      new_units_added
+    FROM
+      libFuzzer_stats.TestcaseRun
+    WHERE
+      ((strategy_mutator_plugin = 0)
+        OR (strategy_mutator_plugin IS NULL))
+      AND DATE_DIFF(CAST(CURRENT_TIMESTAMP() AS DATE), CAST(_PARTITIONTIME AS DATE), DAY) < 31
+      AND ((strategy_corpus_mutations = 0)
+        OR (strategy_corpus_mutations IS NULL))
+      AND ((strategy_handle_unstable = 0)
+        OR (strategy_handle_unstable IS NULL))
+      AND ((strategy_weighted_mutations = 0)
+        OR (strategy_weighted_mutations IS NULL)))
+  GROUP BY
+    strategy)
+"""
+
+
+def _query_multi_armed_bandit_probs(client):
+  """Get query results.
+
+  Queries above query (MULTI_ARMED_BANDIT_PROB) and yields results
```

Comment:
Nit: no need for the parenthetical here.

Reply:
Looks like this and a few others are still open issues. I won't keep commenting on all of them but please do another pass to make sure these are addressed.